File: dungeons-1.21.11/src/main/java/fishmod/features/FishModAddonApi.java
package fishmod.features;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;
import java.util.function.Supplier;

/**
 * Small public extension point so other mods (e.g. FishModAddons) can register a toggle into the
 * Dungeon column of /fm without FishMod depending on them, and without exposing FishModScreen's
 * private Feature/Column types across the mod boundary.
 */
public final class FishModAddonApi {
    private FishModAddonApi() {}

    public record ExternalToggle(String name, String description, Supplier<Boolean> get, Consumer<Boolean> set) {}

    public static final List<ExternalToggle> dungeonToggles = new CopyOnWriteArrayList<>();

    /** Registers a toggle that appears in /fm's Dungeon column, e.g. for an addon's own on/off switch. */
    public static void registerDungeonToggle(String name, String description, Supplier<Boolean> get, Consumer<Boolean> set) {
        dungeonToggles.add(new ExternalToggle(name, description, get, set));
    }
}

File: dungeons-26.1.2/src/main/kotlin/fishmod/features/FishModAddonApi.kt
package fishmod.features

import java.util.concurrent.CopyOnWriteArrayList
import java.util.function.Consumer
import java.util.function.Supplier

object FishModAddonApi {

    class ExternalToggle(
        private val toggleName: String,
        private val toggleDescription: String,
        private val toggleGet: Supplier<Boolean>,
        private val toggleSet: Consumer<Boolean>
    ) {
        fun name(): String = toggleName
        fun description(): String = toggleDescription
        fun get(): Supplier<Boolean> = toggleGet
        fun set(): Consumer<Boolean> = toggleSet
    }

    @JvmField
    val dungeonToggles: MutableList<ExternalToggle> = CopyOnWriteArrayList()

    @JvmStatic
    fun registerDungeonToggle(name: String, description: String, get: Supplier<Boolean>, set: Consumer<Boolean>) {
        dungeonToggles.add(ExternalToggle(name, description, get, set))
    }

    @JvmField
    val cheatToggles: MutableList<ExternalToggle> = CopyOnWriteArrayList()

    @JvmStatic
    fun registerCheatToggle(name: String, description: String, get: Supplier<Boolean>, set: Consumer<Boolean>) {
        cheatToggles.add(ExternalToggle(name, description, get, set))
    }
}
